### PR TITLE
chisel: support in-place ops via MemoryEffectOpInterface

### LINKED ISSUE #8385: [python] Derive in-place / MemWrite operand list from ODS
**Context**
`tools/chisel/chisel/ops.py` keeps a hand-maintained map of TTNN op class -> role names of operands that are mutated in place (ODS marker: `Arg<..., [MemWrite]>`). The chisel executor uses this list to return one extra tensor per provided in-place operand after the SSA results, so callers know which values were updated by the op.

```python
# tools/chisel/chisel/ops.py
CHISEL_INPLACE_OPS: dict[type, tuple[str, ...]] = {
    ttnn.UpdateCacheOp: ("cache",),
    ttnn.PagedUpdateCacheOp: ("cache",),
    ttnn.FillCacheOp: ("cache",),
    ttnn.PagedFillCacheOp: ("cache",),
    ttnn.WriteTensorOp: ("device_tensor",),
    ttnn.BatchNormTrainingOp: ("running_mean", "running_var"),
    ttnn.PointToPointOp: ("optional_output_tensor",),
    ...
}
```

This is fragile: new ops with `MemWrite` operands silently behave wrong in chisel (executor returns the wrong number of tensors, assertion fires only at test time), and the list drifts from the source of truth in ODS as ops evolve.

**Source of truth**
ODS already encodes this via per-operand decorators. Example from a TTNN op definition:
```
let arguments = (ins
    AnyRankedTensor:$input,
    Arg<AnyRankedTensor, "", [MemWrite]>:$cache,
    ...
);
```

`tools/scripts/python_op_schema_codegen.py` already parses `llvm-tblgen --dump-json` output and emits `OP_SCHEMA[<op-name>] = {"operands": (...), "attributes": (...), "results": (...)}`. We can extend it to capture per-arg decorators (specifically `MemWrite`) and expose them as part of the same sidecar.

**Proposed change**
1. Extend `python_op_schema_codegen.py` to record per-operand decorator info. Suggested shape:
   ```python
   OP_SCHEMA["ttnn.update_cache"] = {
       "operands":   ("cache", "input", ...),
       "memwrite_operands": ("cache",),
       "attributes": (...),
       "results":    (...),
   }
   ```
2. Replace `CHISEL_INPLACE_OPS` lookups in `chisel/ops.py` with a wrapper that reads `OP_SCHEMA[op.name].memwrite_operands` (and returns `()` for ops not in the schema).
3. Delete the hand-maintained `CHISEL_INPLACE_OPS` table.

**Related**
- PR introducing the hand-maintained table: tt-mlir #8354.

### PR DESCRIPTION
**Pull request by ndrakulicTT**
## Summary

Closes #8385

Replaces the hand-maintained `_CHISEL_INPLACE_OPS` table with an IR-derived view of in-place mutation, and teaches chisel's per-op flow to validate mutated operands alongside SSA results.

### What changes

- **C++ binding (`python/Util.cpp`):** new `get_write_effect_operand_indices` that returns flat operand indices on which an op declares `MemoryEffects::Write`, or `None` when the op doesn't implement `MemoryEffectOpInterface`.
- **`chisel/ops.py`:** `get_inplace_vals(op)` now derives mutated tensor operands from the binding above instead of the static dict.
  The `_CHISEL_INPLACE_OPS` table and `get_inplace_operands` helper are removed.
- **`chisel/callbacks.py` — `_default_post_op`:** SSA results and in-place mutated operands are validated in a single loop.
  Each enabled numerics mode (isolated / accumulated) produces a numerics record per result and per mutated operand.
  The pool is refreshed for in-place operands when a golden is registered.
- **`chisel/callbacks.py` — `_evict_inplace_no_golden`:** for no-golden ops that declare MemWrite, the pooled golden for each mutated operand is dropped (with a `golden_evicted` record), preventing false PCC failures on downstream ops.
- **`chisel/executor.py`:** drops `execute_golden_from_pool` — the unified post-op loop in `callbacks.py` handles pool refresh inline.
- **`chisel/op_configs.py`:** `UpdateCacheOp`, `PagedUpdateCacheOp`, `FillCacheOp`, `PagedFillCacheOp`, and `BatchNormTrainingOp` no longer need to be marked `no_golden=True` — they have working chisel goldens and run through the in-place validation path.
- **Tests:** new `test_chisel_records_update_cache_inplace` and `test_chisel_records_batch_norm_training_inplace` exercising the new path end-to-end (numerics records under both isolation and accumulation, no spurious golden_evicted records).

## Follow-ups

Two follow-up PRs are planned on top of this one:

1. **Device tensor pool** — reintroduce the program-scoped device-tensor cache that was prototyped on this branch and then removed before merge.
   Goal: stop redundantly pulling the same SSA to host across consumer ops and across iso/accum modes.
2. **Test refactor** — the two new in-place tests in this PR share most of their structure with the existing chisel integration tests (session setup, record filtering, PCC assertions).
   A follow-up will factor the common scaffolding into shared helpers / parametrize across ops.

### CODE REVIEW

File: python/Util.cpp [hunk 1]
```diff
@@ -2,12 +2,50 @@
...
 
+#include "mlir/Interfaces/SideEffectInterfaces.h"
 #include "ttmlir/Bindings/Python/TTMLIRModule.h"
+#include <nanobind/stl/vector.h>
+#include <optional>
 #include <variant>
 
 namespace mlir::ttmlir::python {
 
+static std::optional<std::vector<int64_t>>
```
Comment (azecevic000) [inline python/Util.cpp]: Return type is a bit inconsistent. Function returns `nullopt` if op doesn't implement `MemoryEffectOpInterface`, but returns an empty vector if it does but there isn't any operand with a `MemoryEffects::Write`. I'm not sure if it makes sense to discriminate between these two cases.

**Review by azecevicTT — APPROVED** [review]
`python/Util.cpp` LGTM, one comment inline.

File: tools/chisel/chisel/callbacks.py [hunk 2]
```diff
@@ -152,58 +155,101 @@ def _emit_pcc(
...
-        accum_outs = [None] * len(mlir_op_outputs)
+        modes.append(NumericsMode.ACCUMULATED)
 
-    for mlir_output, output_ref, iso_out, accum_out in zip(
-        mlir_op_outputs, ctx.output_refs, iso_outs, accum_outs, strict=True
-    ):
-        device_tensor = _validate_and_retrieve_tensor(ctx, mlir_output, output_ref)
-        ssa = mlir_output.get_name(asm_state)
+    if len(modes) == 0:
+        return
```
Comment (mmilosevicTT) [inline tools/chisel/chisel/callbacks.py]:
```suggestion
    if not (ctx.checks_config.isolation or ctx.checks_config.accumulated):
        return
    modes: list[NumericsMode] = []
    if ctx.checks_config.isolation:
        modes.append(NumericsMode.ISOLATED)
    if ctx.checks_config.accumulation:
        modes.append(NumericsMode.ACCUMULATED)
```

Or I would even fail if chisel context is run without any mode set.

File: tools/chisel/chisel/callbacks.py [hunk 3]
```diff
@@ -152,58 +155,101 @@ def _emit_pcc(
...
+                mlir_value,
+                golden_out,
+                device_tensors[idx],
+                mode=mode,
                 skip_pcc=config.skip_pcc,
             )
+            if mode is NumericsMode.ISOLATED:
+                continue
+
+            ctx.golden_tensor_pool[ssa] = golden_out
```
Comment (mmilosevicTT) [inline tools/chisel/chisel/callbacks.py:252]:
```suggestion
            if mode is NumericsMode.ACCUMULATED:
                ctx.golden_tensor_pool[ssa] = golden_out
```

**Review by mmilosevicTT — APPROVED** [review]
LGTM, apart from changes that other guys requested

File: tools/chisel/chisel/callbacks.py (hunk 3 above)
Comment (ndrakulicTT) [reply tools/chisel/chisel/callbacks.py:252]: Leaving this for now, as next PRs will only have logic for the accumulated mode
